aten/src/ATen/native/vulkan/Vulkan.cpp

```diff
@@ -896,6 +908,79 @@ VBuffer makeUniformConstBuffer(void* ptr, VkDeviceSize size) {
   return constBuffer;
 }
 
+ComputeUnitFactory::ComputeUnitFactory(const VkDevice& device)
+    : device_(device) {
+  VkPipelineCacheCreateInfo createInfo{};
+  createInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_CACHE_CREATE_INFO;
+  createInfo.pNext = nullptr;
+  createInfo.flags = 0;
+  createInfo.initialDataSize = 0;
+  createInfo.pInitialData = nullptr;
+  VK_CHECK(vkCreatePipelineCache(
+      device_, &createInfo, nullptr /* allocator */, &pipelineCache_));
+}
+
+ComputeUnitFactory::~ComputeUnitFactory() {
+  vkDestroyPipelineCache(device_, pipelineCache_, nullptr /* allocator */);
+}
+
+std::string ComputeUnitFactory::getCacheKey(
+    const char* key,
+    WorkGroupSize& workGroupSize) {
+  std::stringstream ss;
+  ss << key << ':' << workGroupSize.x << ':' << workGroupSize.y << ':'
+     << workGroupSize.z;
+  return ss.str();
+}
+
+ComputeUnit& ComputeUnitFactory::get(
+    const std::string& cacheKey,
+    std::function<std::shared_ptr<ComputeUnit>()> factoryFn) {
+  auto it = computeUnits_.find(cacheKey);
+  if (it != computeUnits_.end()) {
+    return *(it->second.get());
+  }
+  auto computeUnit = factoryFn();
+  computeUnits_.insert(std::make_pair(cacheKey, computeUnit));
+  return *(computeUnit.get());
+}
+
+#ifdef USE_VULKAN_SHADERC_RUNTIME
```

Comment:
One alternate design could have been to encapsulate the difference between source code (null-terminated const char* shader_source) and binary representation ([const uint32_t* code,  const uint32_t size] pair) such that the need for multiple code paths could have been avoided.  Haven't put much thought into it but I think it should be possible to cheaply abstract that away.